### PR TITLE
RPC enhancements: scaffolding, notification broadcasting, gRPC client & server

### PR DESCRIPTION
**Pull request by tiram88**
- further scaffolding and implementing of the RPC sub-system
- make the whole notification system generic over a Connection trait
- optimize the notification broadcasting by preemptively converting notifications into messages of the type and variants expected by the listeners
- refactoring of the gRPC client
- add message ids to gRPC server and client to get unambiguous request/response matching
- gRPC client adapts to server features regarding support for message ids and notification unsubscription
- add an auto-reconnect feature to the gRPC client
- put gRPC client and server in separate crates
- rename crates:
   - rpc-core   =>   kaspa-rpc-core
   - rpc-grpc   =>   kaspa-grpc-core + kaspa-grpc-client + kaspa-grpc-server

### CODE REVIEW

File: rpc/core/src/notify/connection.rs [hunk 1]
```diff
@@ -0,0 +1,61 @@
+use crate::{Notification, NotificationSender};
+use std::fmt::Debug;
+use std::{hash::Hash, sync::Arc};
+
+pub trait Connection: Clone + Debug + Send + Sync + 'static {
+    type Message: Clone;
+    type Variant: Hash + Clone + Eq + PartialEq + Send;
```
[the remark this answers is not in the document]
Comment (tiram88) [reply rpc/core/src/notify/connection.rs]: Renamed to Encoding

File: rpc/core/src/api/notifications.rs [hunk 2]
```diff
@@ -24,12 +23,12 @@ impl From<&Notification> for NotificationType {
         match item {
             Notification::BlockAdded(_) => NotificationType::BlockAdded,
             Notification::VirtualSelectedParentChainChanged(_) => NotificationType::VirtualSelectedParentChainChanged,
-            Notification::FinalityConflict(_) => NotificationType::FinalityConflicts,
+            Notification::FinalityConflict(_) => NotificationType::FinalityConflict,
             Notification::FinalityConflictResolved(_) => NotificationType::FinalityConflictResolved,
             Notification::UtxosChanged(_) => NotificationType::UtxosChanged(vec![]),
```
[the remark this answers is not in the document]
Comment (tiram88) [reply rpc/core/src/api/notifications.rs:28]: The lossy conversion from `Notification` for `NotificationType` is no longer needed.

Removed.

File: rpc/core/src/api/Extending RpcApi.md [hunk 3]
```diff
@@ -32,12 +32,12 @@ As an illustration, let's pretend that we add a new `submit_block` method.
 2. In file `rpc\grpc\proto\messages.proto`, add respectively a request and a response to the payload of `KaspadRequest` and `KaspadResponse`.
    (ie. `SubmitBlockRequestMessage submitBlockRequest = 1003;` and `SubmitBlockResponseMessage submitBlockResponse = 1004;`)
 3. In `rpc\grpc\src\convert\message.rs`, implement converters to handle rpc-core <-> rpc-grpc.
-4. If appropriate, implement a matcher in `rpc_grpc::client::resolver::matcher`.
+4. If appropriate, implement a matcher in `kaspa_rpc_grpc::client::resolver::matcher`.
```
[the remark this answers is not in the document]
Comment (tiram88) [reply rpc/core/src/api/Extending RpcApi.md:35]: fixed

File: rpc/core/src/model/tx.rs [hunk 4]
```diff
@@ -82,12 +81,3 @@ pub struct RpcAcceptedTransactionIds {
     pub accepting_block_hash: RpcHash,
     pub accepted_transaction_ids: Vec<TransactionId>,
```
[the remark this answers is not in the document]
Comment (tiram88) [reply rpc/core/src/model/tx.rs:82]: fixed

File: rpc/core/src/api/Extending RpcApi.md [hunk 5]
```diff
@@ -32,12 +32,12 @@ As an illustration, let's pretend that we add a new `submit_block` method.
...
 6. In `rpc\grpc\src\convert\kaspad.rs`, complete the `From` implementations for `RpcApiOps`.
 7. In `rpc\grpc\src\convert\kaspad.rs`, add calls to `impl_into_kaspad_request!` and `impl_into_kaspad_response!`
-   (ie. `impl_into_kaspad_request!(rpc_core::SubmitBlockRequest, SubmitBlockRequestMessage, SubmitBlockRequest);` and
-   `impl_into_kaspad_response!(rpc_core::SubmitBlockResponse, SubmitBlockResponseMessage, SubmitBlockResponse);`).
-8. Implement the function having a `_call` suffix into `rpc_grpc::client::RpcApiGrpc`.
-9. In `rpc_grpc::server::service::RpcService::message_stream`, requests handler, add an arm and implement
+   (ie. `impl_into_kaspad_request!(kaspa_rpc_core::SubmitBlockRequest, SubmitBlockRequestMessage, SubmitBlockRequest);` and
+   `impl_into_kaspad_response!(kaspa_rpc_core::SubmitBlockResponse, SubmitBlockResponseMessage, SubmitBlockResponse);`).
+8. Implement the function having a `_call` suffix into `kaspa_rpc_grpc::client::RpcApiGrpc`.
+9. In `kaspa_rpc_grpc::server::service::RpcService::message_stream`, requests handler, add an arm and implement
```
[the remark this answers is not in the document]
Comment (tiram88) [reply rpc/core/src/api/Extending RpcApi.md:42]: fixed

**Review by michaelsutton — APPROVED** [review]

**Pull request merged**